stan/math/rev/functor/adj_jac_apply.hpp

```diff
@@ -405,16 +325,16 @@ struct adj_jac_vari : public vari {
    * @param args the rest of the arguments (that will be iterated through
    * recursively)
    */
-  template <int R, int C, typename... Pargs>
-  inline void accumulate_adjoints(const Eigen::Matrix<double, R, C>& y_adj_jac,
-                                  const Pargs&... args) {
+  template <typename T, typename... Pargs,
+            require_eigen_st<std::is_arithmetic, T>...>
+  inline void accumulate_adjoints(T&& y_adj_jac,
+                                  Pargs&&... args) {
     static constexpr int t = sizeof...(Targs) - sizeof...(Pargs) - 1;
     if (is_var_[t]) {
-      for (int n = 0; n < y_adj_jac.size(); ++n) {
-        x_vis_[offsets_[t] + n]->adj_ += y_adj_jac(n);
-      }
+      Eigen::Map<matrix_vi>(&x_vis_[offsets_[t]], y_adj_jac.rows(),
+                            y_adj_jac.cols()).adj() += std::move(y_adj_jac);
```

Comment:
I'd pull out the rows and cols before doing anything to make sure the object's elements exist when this is all executed
```suggestion
      const y_adj_jac_cols = y_adj_jac.cols()
      const y_adj_jac_rows = y_adj_jac.rows()
      Eigen::Map<matrix_vi>(&x_vis_[offsets_[t]], y_adj_jac_rows,
                            y_adj_jac_cols).adj() += std::forward<T>(y_adj_jac);
```